File: api/lib/signature_guard.py
"""
Signature Guard - Project Authentication System
Author: kk
DO NOT REMOVE - Critical for project operation
"""
import os
import sys
import hashlib
from pathlib import Path


class SignatureGuard:
    """Project signature verification and authentication system"""
    
    _instance = None
    _verified = False
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(SignatureGuard, cls).__new__(cls)
        return cls._instance
    
    def __init__(self):
        self.signature_file = Path(__file__).parent.parent.parent / ".signature"
        self.required_keys = ["DEVELOPER", "PROJECT_ID", "BUILD_SIGNATURE", "LICENSE_KEY", "INTEGRITY_HASH"]
        self.signature_data = {}
    
    def _load_signature(self) -> bool:
        """Load and parse signature file"""
        try:
            if not self.signature_file.exists():
                return False
            
            with open(self.signature_file, 'r') as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith('#'):
                        if '=' in line:
                            key, value = line.split('=', 1)
                            self.signature_data[key] = value
            
            return all(key in self.signature_data for key in self.required_keys)
        except Exception:
            return False
    
    def _verify_integrity(self) -> bool:
        """Verify signature integrity"""
        try:
            expected_values = [
                self.signature_data.get("DEVELOPER", ""),
                self.signature_data.get("PROJECT_ID", ""),
                self.signature_data.get("BUILD_SIGNATURE", "")
            ]
            
            # Check all required components are present
            if not all(expected_values):
                return False
            
            # Verify specific signature patterns
            if "kk" not in self.signature_data.get("DEVELOPER", ""):
                return False
            
            if "MINI" not in self.signature_data.get("PROJECT_ID", ""):
                return False
            
            if "KK" not in self.signature_data.get("BUILD_SIGNATURE", ""):
                return False
            
            return True
        except Exception:
            return False
    
    def verify(self) -> bool:
        """Main verification method - DO NOT REMOVE"""
        if self._verified:
            return True
        
        # Load signature
        if not self._load_signature():
            self._handle_verification_failure("Signature file missing or corrupted")
            return False
        
        # Verify integrity
        if not self._verify_integrity():
            self._handle_verification_failure("Signature verification failed")
            return False
        
        self._verified = True
        return True
    
    def _handle_verification_failure(self, reason: str):
        """Handle verification failure"""
        error_msg = f"""
╔══════════════════════════════════════════════════════════════╗
║           PROJECT SIGNATURE VERIFICATION FAILED              ║
╠══════════════════════════════════════════════════════════════╣
║                                                              ║
║  This project is protected and authenticated.               ║
║  The signature file has been removed or tampered with.      ║
║                                                              ║
║  Original Developer: kk                                     ║
║  Project: MINI AI Database Assistant                        ║
║                                                              ║
║  Reason: {reason:<50} ║
║                                                              ║
║  Please contact the original developer for a valid license. ║
║                                                              ║
╚══════════════════════════════════════════════════════════════╝
"""
        print(error_msg, file=sys.stderr)
        sys.exit(1)
    
    def get_developer_info(self) -> dict:
        """Get developer information"""
        if not self._verified:
            self.verify()
        
        return {
            "developer": self.signature_data.get("DEVELOPER", "Unknown"),
            "project": self.signature_data.get("PROJECT_ID", "Unknown"),
            "signature": self.signature_data.get("BUILD_SIGNATURE", "Unknown")
        }


# Global instance
_guard = SignatureGuard()


def verify_signature() -> bool:
    """Verify project signature - Critical function"""
    return _guard.verify()


def get_developer_info() -> dict:
    """Get developer information"""
    return _guard.get_developer_info()


# Auto-verify on import - DO NOT REMOVE
if not verify_signature():
    sys.exit(1)

File: frontend/src/utils/signatureGuard.js
/**
 * Signature Guard - Project Authentication System
 * Author: kk
 * DO NOT REMOVE - Critical for project operation
 */

class SignatureGuard {
  constructor() {
    this.verified = false;
    this.signature = {
      developer: 'kk',
      project: 'MINI_AI_DATABASE_ASSISTANT_FRONTEND',
      buildSignature: 'KK_2026_MINI_AUTH_PROTECTED_UI',
      licenseKey: 'ui_7d8a9c4b_frontend_kk_2026_protected'
    };
  }

  /**
   * Verify project signature - DO NOT REMOVE
   */
  verify() {
    if (this.verified) {
      return true;
    }

    try {
      // Check signature integrity
      const isValid = this._verifyIntegrity();
      
      if (!isValid) {
        this._handleVerificationFailure('Signature verification failed');
        return false;
      }

      this.verified = true;
      return true;
    } catch (error) {
      this._handleVerificationFailure('Signature check error');
      return false;
    }
  }

  /**
   * Verify signature integrity
   */
  _verifyIntegrity() {
    // Check all required components
    if (!this.signature.developer || !this.signature.project || 
        !this.signature.buildSignature || !this.signature.licenseKey) {
      return false;
    }

    // Verify specific patterns
    if (!this.signature.developer.includes('kk')) {
      return false;
    }

    if (!this.signature.project.includes('MINI')) {
      return false;
    }

    if (!this.signature.buildSignature.includes('KK')) {
      return false;
    }

    return true;
  }

  /**
   * Handle verification failure
   */
  _handleVerificationFailure(reason) {
    const errorMsg = `
╔══════════════════════════════════════════════════════════════╗
║           PROJECT SIGNATURE VERIFICATION FAILED              ║
╠══════════════════════════════════════════════════════════════╣
║                                                              ║
║  This project is protected and authenticated.               ║
║  The signature has been removed or tampered with.           ║
║                                                              ║
║  Original Developer: kk                                     ║
║  Project: MINI AI Database Assistant                        ║
║                                                              ║
║  Reason: ${reason.padEnd(50)} ║
║                                                              ║
║  Please contact the original developer for a valid license. ║
║                                                              ║
╚══════════════════════════════════════════════════════════════╝
    `;
    
    console.error(errorMsg);
    
    // Display error to user
    document.body.innerHTML = `
      <div style="display: flex; align-items: center; justify-content: center; height: 100vh; background: #0f0f0f; color: #fff; font-family: monospace;">
        <div style="background: #1a1a1a; border: 2px solid #ff4444; border-radius: 8px; padding: 40px; max-width: 600px; text-align: center;">
          <h1 style="color: #ff4444; margin-bottom: 20px;">⚠️ Signature Verification Failed</h1>
          <p style="margin-bottom: 20px; line-height: 1.6;">
            This project is protected and authenticated.<br/>
            The signature has been removed or tampered with.
          </p>
          <div style="background: #0f0f0f; padding: 20px; border-radius: 4px; margin-bottom: 20px;">
            <p style="margin: 5px 0;"><strong>Original Developer:</strong> kk</p>
            <p style="margin: 5px 0;"><strong>Project:</strong> MINI AI Database Assistant</p>
            <p style="margin: 5px 0;"><strong>Build:</strong> KK_2026_MINI_AUTH_PROTECTED</p>
          </div>
          <p style="color: #ff4444;">Reason: ${reason}</p>
          <p style="margin-top: 20px; font-size: 12px; color: #888;">
            Please contact the original developer for a valid license.
          </p>
        </div>
      </div>
    `;
    
    throw new Error('Signature verification failed');
  }

  /**
   * Get developer information
   */
  getDeveloperInfo() {
    if (!this.verified) {
      this.verify();
    }
    return {
      developer: this.signature.developer,
      project: this.signature.project,
      signature: this.signature.buildSignature
    };
  }
}

// Create global instance
const signatureGuard = new SignatureGuard();

// Auto-verify on load - DO NOT REMOVE
if (!signatureGuard.verify()) {
  throw new Error('Signature verification failed');
}

export default signatureGuard;
export const verifySignature = () => signatureGuard.verify();
export const getDeveloperInfo = () => signatureGuard.getDeveloperInfo();
